core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -622,32 +622,26 @@ public TableMetadata buildReplacement(Schema updatedSchema, PartitionSpec update
     ValidationException.check(formatVersion > 1 || PartitionSpec.hasSequentialIds(updatedPartitionSpec),
         "Spec does not use sequential IDs that are required in v1: %s", updatedPartitionSpec);
 
-    AtomicInteger nextLastColumnId = new AtomicInteger(0);
-    Schema freshSchema = TypeUtil.assignFreshIds(updatedSchema, nextLastColumnId::incrementAndGet);
+    AtomicInteger newLastColumnId = new AtomicInteger(lastColumnId);
+    Schema freshSchema = TypeUtil.assignFreshIds(updatedSchema, schema, newLastColumnId::incrementAndGet);
 
-    int nextSpecId = TableMetadata.INITIAL_SPEC_ID;
-    for (Integer specId : specsById.keySet()) {
-      if (nextSpecId <= specId) {
-        nextSpecId = specId + 1;
-      }
-    }
+    // determine the next spec id
+    OptionalInt maxSpecId = specs.stream().mapToInt(PartitionSpec::specId).max();
+    int nextSpecId = maxSpecId.orElse(TableMetadata.INITIAL_SPEC_ID) + 1;
```

Comment:
I don't think it matters but it seems the computations for `nextSpecId` and `nextOrderId` are a bit different now. Previously, if partition spec list was empty, we used  `INITIAL_SPEC_ID` and if sort order list was empty, we used `INITIAL_SORT_ORDER_ID `. Now we are going to use `INITIAL_SPEC_ID + 1` in that case.

Reply:
Yes, this simplifies the logic a bit and avoids using spec id 0.